pkg/controllers/nodeclaim/capacityreservation/controller.go

```diff
@@ -0,0 +1,83 @@
+package capacityreservation
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+)
+
+type Controller struct {
+	cp cloudprovider.CloudProvider
+	kubeClient client.Client
+}
+
+func NewController() *Controller {
+	return nil
+}
+
+func (*Controller) Name() string {
+	return "nodeclaim.capacityreservation"
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named(c.Name()).
+		WatchesRawSource(singleton.Source()).
+		Complete(singleton.AsReconciler(c))
+}
+
+func (c *Controller) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	cpNodeClaims, err := c.cp.List(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+	}
+	cpNodeClaimIndex := lo.SliceToMap(cpNodeClaims, func(nc *karpv1.NodeClaim) (string, *karpv1.NodeClaim) {
+		return nc.Status.ProviderID, nc
+	})
+
+
+	var ncs *karpv1.NodeClaimList
+	if err := c.kubeClient.List(ctx, ncs); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodeclaims, %w", err)
+	}
+	var errs []error
+	for i := range ncs.Items {
+		cpNC, ok := cpNodeClaimIndex[ncs.Items[i].Status.ProviderID]
+		if !ok {
+			continue
+		}
+		// We only want to react if a current reserservation has been demoted to on-demand, and we only care if we wer
+		// tracking it previously. This ensures customers who were previously utilizing open ODCRs don't encounter node
+		// selection issues due to the labels being migrated.
+		if ncs.Items[i].Labels[karpv1.CapacityTypeLabelKey] == karpv1.CapacityTypeReserved || cpNC.Labels[karpv1.CapacityTypeLabelKey] != karpv1.CapacityTypeReserved {
```

Comment:
I'm not sure that I follow this check -- this says that if the current NodeClaim is marked as reserved OR the cloudprovider NodeClaim isn't marked as reserved then we should continue? Don't we want the opposite? If the NodeClaim hasn't previously been marked as reserved OR the CloudProvider NodeClaim isn't marked as reserved, those would be the scenarios I would expect that we would want to alter the labels

Reply:
Updated in the current iteration, this check was incorrect and should have been an AND. Either way though, the current implementation has been refactored quite a bit.